[Audio] One PR to rule them all, One PR to find them, One PR to bring them all, and in the darkness bind them (all-in-one pr)

Drapersniper opened the pull request:
### Type

- [x] Bugfix
- [x] Enhancement
- [x] New feature

### Description of the changes

After discussing with @aikaterna she prefers everything in one PR so that conflicts and reviewing cross interaction and testing can be done more easily.

This PR will superseded #2861 and #2890, in addition to those 2 I've done a lot more work on other key areas of audio.

- **Bug Fixes**

1. Playlist save fix (will overwrite playlists with no confirmation)
    - Playlist are now stored by ID (`ctx.message.id`), this allows multiple playlist with the same name
2. `[p]playlist remove` now converts a playlist to a custom one if it was created from a link
    - Before this part of the code was always ignored
3. loadtrack keyerror fixed (#2711)
4. Stop showing lavalink jar version in `[p]audioset settings` fixed ( #2880 )

- **Done**

1. Moved Playlist to a custom scope, there are 3 new custom scopes
    - `Global` Which only Bot Owner have edit access
    - `Guild` Which bot owners, DJ, Mods, Admins, Guild Owners, Authors have access to
    - `User` which bot owner and playlist author have access
2.  Playlists are now stored in `Playlist` object
3. Created API's to easily access/modify/delete Playlists based on the ModLog structure
4. Wrote a function that converts the old config schema to the new versions with the 
5. Update all playlist commands to use the new `Playlist` objects
6. Rename Playlists command `[p]playlist rename` added
7. V3 now have their own playlist schema that has song information, so when a playlist is uploaded no song information will need to be re-fetched, keeping support for V2 playlists
8. Making a dynamic converter for the Playlist ID, so that users can give the bot either the Playlist ID or the playlist name, it will prompt user to select one if multiple matches are found
9. Added `[p]playlist update` (If a playlist has a playlist url saved (non-custom playlist)) will now update the playlist to the latest tracks available.
10.  Child folders within "album" folders in localtracks (#270)
11. Implemented `LocalPath` and `Query` data classes to handle queries and paths.
12. Playlist now start playing after first song is loaded instead of waiting for the full playlist to load
13. Lavalink response is now cacheable as part of #2890 (Significantly faster loading, 2.7K songs went from 1+ hours to 1-2 minutes on follow up enqueues)
14. As part of the changes added by  #2890 Users are no longer able to enqueue any tracks while a playlist is enqueuing, this is to ensure queue order persists. But since songs now start being played the moment they are loaded, This doesn't have a major negative impact to users.
15. Dispatches for on_track_start, on_track_end, on_queue_end, on_track_enqueue, on_skip_track, on_auto_play_track
16. Users can turn on autoplay so songs automatically play when the queue is empty
17. Users can set one of their playlist to be used for auto play 
18. Adding and removing reactions are no longer blocking actions
19. YouTube and Spotify API calls are also cacheable
20. Cache for each level can be toggled individually
21. Local tracks folder can be anywhere in your machine as long as your lavalink has perms to see the tracks.
22. Rework localpath command to support new localpath logic
23. Add command to Set cache age 
24. Make auto play playlist customizable
25. url or keyword blacklist  and whitelist
26. Toggle for autopause when no one is in the channel (#721)
27. `[p]genre` allows you to queue a Spotify playlist from a set of categories
28. `[p]playlist update` shows the changes done to the playlist

## Requires https://github.com/Cog-Creators/Red-Lavalink/pull/61 as a dependency
1. Full timestamp support added regardless of when the track is added
2. Timestamp also work when searching for songs, if you use the services timestamp
3. Queue order is fully correct when shuffle is on
4. Allow skip with shuffle on
5. Track Enqueue embeds now give the correct track number if shuffle is on
6. Users can manually shuffle the queue with `[p]queue shuffle`

Drapersniper committed:
More changes

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed auto play defaulting to playlist

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Localtrack fix

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Updated deps .. since for some reason aiosqlite is not being auto installed for everyone

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Yupo

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Fixed a crash in [p]now

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed crash on playlist save

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Debugging Commit

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Yet more prints

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Even more spammy debug

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Debugging commit + NEw Dispatches

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Debugging commit + NEw Dispatches

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed localpath checks

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
more fixes for Localpaths

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Spelling mistake on method

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed Crash on event handler

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed Crash on local search

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Reduced fuzzy match percentage threshold for local tracks to account for nested folders

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed a crash on queue end

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Sigh ... Removed a duplicate dispatch

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Sigh i removed this before ...

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Reorder dispatch signatures so all 3 new dispatch have matching signature

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Formatting

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Edited Error Event to support localtracks

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix a Crash on track crash :awesome:

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Yikes soo much spam

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Remove spam and improve existance check

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Repeat and Auto-play are mutually exclusive now

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
DEBUGS for Preda

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Vimeo tracks can be from both these domains "vimeo.com", "beam.pro"

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
I mean Mixer can be from those 2 domains ....

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed `search sc` command

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Run everything though lints.
rename localtracks module to dataclasses
Clear lock on errors

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Try to speed up long playlist loading

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Im an idiot

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Im an idiot

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Added logging for writes

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fix crash on cog reload

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fix for runtimewarning ?

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fix for Local Track cache

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Remove broken tracks from queue on exception
Theoretically do not auto play if track stop reason is Stopped or cleanup

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Previous commit was a fluke ... ignore it

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Change from cleanup to Replaced

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fixed AttributeError: 'Track' object has no attribute 'info'.
[p]skip will only work for autoplay is there a track being played.
Fixed Console spam if query saving failed in the background while reloading bot.
Autoplay now respect [p]stop command

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black formatting
Fix Issue with auto play working when there is songs in the queue
Stop notifying queue ended if autoplay is on

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed a crash on track load timeout

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
[p]playlist start will now show the playlist name in embed body
Improved Logic for handling broken tracks when repeat is on.

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Enqueue tracks as soon as we have the youtube URL ....

This basically changes how spotify urls are handled

Need to test saving spotify playlist
Need to test loading a spotify playlist from file
Need to test enqueuing a spotify playlist

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Updated a track whrn enqueuing spotify playlist

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Debug

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Revert spotify_enqueue changes

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Merge remote-tracking branch 'origin/audiochanges' into audiochanges

Drapersniper committed:
Revert spotify_enqueue changes

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Allow to set Lavalink jar version from Environment vars

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Allow to set Lavalink jar version from Environment vars

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fix for a crash on Equalizer, Merge Spotify_enqueue changes and revert manager changes

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Break playlist enqueue after 10 consecutive failures

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Auto DC, is not compatible with Auto Play

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Make notifier aware of guild its being called for

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Type checking

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Remove lock from 2 exits that i didn't before

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fixed TypeError: spotify_enqueue() got an unexpected keyword argument 'notify'

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Reorder toggles to alphabetical order

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Update Query to handle spotify URIs

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
update database

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Dont say tracks enqued on invalid link
Make autop lay a mod only setting

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Merge remote-tracking branch 'origin/audiochanges' into audiochanges

Drapersniper commented:
So I think it may be ready for review ....
I'm truly sorry for all the reviewers and QA's ... On the up side the code works ... on the downside i probably missed a fair amount of docs if you notice anything obvious don't hesitate to shout

Drapersniper committed:
Dont say tracks enqued on invalid spotify link

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Set default age to 365 days

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Allow Audio mods to set auto play playlists.
Save playlists songs to cache when migrating

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black formatting

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
[p]eq cooldown is not triggered is player check fails (i.e if nothing is currently playing)
Adding and removing reaction is no longer a blocking action

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
changelog for non blocking reaction handles

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Show auto dc  and auto play settings by default

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
lint is being a bitch

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
lint changes

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
stop caching local tracks

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
List of Lavalink.Tracks natively added to Playlist Objects

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fix UX changes and should fix autoplay

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Fixed Skip x number of tracks

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Lint changes

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Remvoe dead code

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Update playlist embed formatting to reflect Preda's suggestions

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Update change logs

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Add `async with ctx.typing():` to queue and to local folder

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Stop queuing now when queue is empty with [p]queue

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
fix ctx.typing()

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
fix ctx.typing()

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Part 1

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Dont check local track author and name if title is Unknown

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Makes auto play more random

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes local play
Fixed missing format

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Query.process_input accept lavalink.Track objects

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
docstrings

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Add TODO for timestamp support

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Improve autoplay from cache logic (possibly slightly slower but more efficient overall)

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Add My Lavalink PR as a dependency
Remember to remove this .... The PR will bump it to 0.3.2

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Add My Lavalink PR as a dependency
Remember to remove this .... The PR will bump it to 0.3.2

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Add My Lavalink PR as a dependency
Remember to remove this .... The PR will bump it to 0.3.2

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Compile all regex at runtime

Signed-off-by: Draper <guyreis96@gmail.com>

mikeshardmind commented:
if you need.changes for red-lavalink, PR those and cross link it here. We cant handle the red-lavalink dep the way you have it here.

Drapersniper commented:
> if you need.changes for red-lavalink, PR those and cross link it here. We cant handle the red-lavalink dep the way you have it here.

Yeah I know this was just for testing I fully intend to revert it to the correct way before this is merged

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Fixes local play
Fixed missing format

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Merge remote-tracking branch 'origin/audiochanges' into audiochanges

Drapersniper committed:
Revert Dep error

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed attribute error

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
add `self.bot.dispatch("audio_disconnect", ctx.guild)` dispatch when the player is disconnected

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Removed shuffle lock on skip

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Better logic for auto seek (timestamps)

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Better logic for auto seek (timestamps)

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes timestamps on spotify tracks

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Add ctx typing to playlist enqueue

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix Deps

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black formatting + Using new lavalink methods for shuffling

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
remove ctx.typing from playlist start

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes typerror when enqueuing spotify playlists

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix keyerror

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
black formatting, + embed for [p]audioset cache as I forgot it before

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix Error on playlist upload

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix Text help for bump

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Allow track bumping while shuffle is on

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Edit bump embed to be consistent with other embed
Hyperlink tracks and removed dynamic title

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Errors not printing fix?

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Errors not printing fix?

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Track enqueued footer now shows correct track position when shuffle is on

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Update changelogs

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix is_owner check in audioset settings

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Changelogs

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Dont store searches with no results in cache, fix malformated playlist to cache upon settings migration

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
_clear_lock_on_error > Needs to be reviewed to see if it has been done correctly

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
_clear_lock_on_error > Needs to be reviewed to see if it has been done correctly

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix Query search so that it works with absolute paths for localtracks

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Extra error if lavalink is set to external and  the query is a localtrack and nothing is found

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
More detailed error message

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
[p]seek and [p]skip can be used by user if they are the song requester while DJ mode is enabled, if votes are disabled. , [p]queue shuffle can be used to shuffle the queue manually. and [p]queue clean self can be used to remove all songs you requested from the queue.

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
All the fixes + a `should_auto_play` dispatch for the tech savy peeps

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Spellchecker + Pythonic changes

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
NO spam for logs

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Pass Current voice channel to `red_audio_should_auto_play` dispatch

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
playlist upload also updates cache in the background

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
playlist upload also updates cache in the background

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Add scope to playlist picker

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Delete Playlist picker message once something is selected

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
OCD Fix

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Facepalm

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix a Potential crash

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Update my stupidity

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Auto Pause +  Skip tracks already in playlist upon playlist append + a command to remove duplicated tracks from playlist

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix DJ mode when Role is deleted - Credits go to Neuro Assassin#4779
Fix an issue where auto play MAY not trigger

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Change log to  Neuro Assassin#4779 fix

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Dont auto pause manual pauses

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Adds `[p]autoplay` that can be run by mods or higher

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
:facepalm:

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
2x :facepalm:

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed wrong import

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Added Autoplay notify

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Added Autoplay notify

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Store Track object as prev song instead of URI

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black why do u hate me

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix command name

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix Autoplay notify

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix missing await and TypeError, Thanks Flame

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Add a list of tracks to show as a menu

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
adds the `[p]genre` command which uses the Spotify and Youtube API

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Enqueue Playlists from genre command

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Pretify `[p]genre`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix a Typo and correct jukebox charge order

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Add genre command to error handling

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Type checking

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Update naming scheme for `[p]genre`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black why do you hate me

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixed `[p]local start`
Playlist picker auto selects if theres just 1 playlist found
`[p]queue cleanself` added

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
*sigh* back compatibility with old localtrack paths

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
*sigh* back compatibility with old localtrack paths, even more

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
*sigh* back compatibility with old localtrack paths Even more

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes localtracks in playlist info command

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Debug Local Strings

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Debug Local Strings

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes `[p]playlist info` for local tracks + fixed error in `[p]remove`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes formatting in `[p]playlist info`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix an issue with User Scope playlists were not being deleted

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Typechecking

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Black

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix the logic of `delegate_autoplay`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix a Crash on Load due to type hinting

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix a Crash on Load due to type hintingBlack + fix order of `red_audio_should_auto_play`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Add `red_audio_initialized` dispatch so that ownership of auto play can be maintained after a reload

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Check if the current owner is loaded before raising an error

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fixes the Existence Check in `delegate_autoplay`

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Turns `own_autoplay` in a property of Audio and improves `delegate_autoplay` Thanks Sinbad!

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Fix for Localtracks playlists

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
When disconnecting send `Disconnecting...`
Fix Stop after a skip
Fix UX discrepancy on Playlist IDs
Fixed Exception when theres a track error

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
add `on_red_audio_unload` dispatch

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fix a crash on track start where `player.current` can be none?

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Missing new line

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Allow `--author` for playlist to be used to filter playlist for an specific author.
Plus a few bugfixes for UX

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Rename `remdupe` to `dedupe`
Make global scope always be referenced as Global
add missing backwards quotes around the Playlist ID for 1 string

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Towncrier entries for dep changes

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Remove track index when shuffle is on
Fix Progress bar for livestreams

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Trigger autoplay on `QUEUE_END` event instead of `TRACK_END`

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Can't reproduce Ians bug but here a safeguard agaisnt it just in case

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fixes 2 Messages that had the wrong formatting

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
standerdize playlist naming scheme

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fix `[p]autoplay` message when Notify is enabled

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
y u h8 me black

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fix an issue with `[p]audioset localpath` where the localtracks folder was incorrect

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Pythonic formatting

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Ugh

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fix a typo

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Silently try to delete messages + fixes error Ian found with `[p]genre`

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
sigh black

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' of https://github.com/Cog-Creators/Red-DiscordBot into audiochanges

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Add humanize_number usage correctly

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Bump RLL to 0.4.0

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Update changelog entries

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Change `bot.db` to new API's added by #2967

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Additional reformatting

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' into audiochanges

Drapersniper committed:
Merge branch 'V3/develop' of https://github.com/Cog-Creators/Red-DiscordBot into audiochanges

Drapersniper committed:
Remove PyCharm noise + Fixes a few Pycharm warnings

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Rework `index` parsing for youtube urls

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

aikaterna reviewed (changes requested):
——
Run playlist download (there were multiple server playlists named the same thing as the query)
Get the file
The timeout replies with “Too many matches found and you did not select which one you wanted.” anyway, when it should return when the file is sent
——
Used the generated, downloaded file in playlist upload
Received:
```
[2019-09-26 10:12:34] [ERROR] red: ValueError
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/discord/ext/commands/core.py", line 79, in wrapped
    ret = await coro(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/redbot/cogs/audio/audio.py", line 4782, in _playlist_upload
    scope_data=(scope, author, guild),
  File "/usr/local/lib/python3.7/site-packages/discord/ext/commands/context.py", line 132, in invoke
    ret = await command.callback(*arguments, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/redbot/cogs/audio/audio.py", line 4353, in _playlist_save
    scope, author, guild, specified_user = scope_data
ValueError: not enough values to unpack (expected 4, got 3)
```
——
Use playlist append with the DJ role on and use a user that does not have the DJ role
The reply responds with “You do not have the permissions to manage playlist owned by {username of the person who did the command}”
The reply should respond with the owner of the playlist instead

Playlist delete has this same issue
Playlist remove has this same issue
——
Have a pleb user queue a playlist so that they are the requester.
Turn on DJ role.
Use [p]skip on the pleb act

```
[2019-09-26 10:43:06] [ERROR] red: Exception in command 'skip'
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/discord/ext/commands/core.py", line 79, in wrapped
    ret = await coro(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/redbot/cogs/audio/audio.py", line 6150, in skip
    if is_requester and not can_skip and skip_to_track is None and skip_to_track > 1:
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```
——
The “Playlist modified” embed header for playlist append when the appending playlist has already-saved tracks should be in title case to match other Audio embed titles.
——
Where is the auto pause setting referred to in 721.feature.rst?
——
So this review isn’t entirely asking for changes, just wanted to mention the genre command is really slick and a great addition.

Drapersniper committed:
Addess Aika's review

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' of https://github.com/Cog-Creators/Red-DiscordBot into audiochanges

Drapersniper committed:
Fix a potential crash, saves guild ID to playlists to avoid an scheme change in the future

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Flame442 commented on changelog.d/audio/2904.enhance.2.rst:
```diff
@@ -0,0 +1 @@
+``[p]queue`` now works where there no tracks in the queue (it shows current track playing).
```
Should be "..where there **are** no tracks.."

Flame442 commented on changelog.d/audio/2904.enhance.6.rst:
```diff
@@ -0,0 +1 @@
+``[p]seek`` and `[p]skip`` can be used by user if they are the song requester while DJ mode is enabled, if votes are disabled.
```
[p]skip is missing a backtick

Drapersniper committed:
Add handling for Python installs without sqlite3.

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Address Flame's review

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fix ma stupidity

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

aikaterna commented:
Playlist rename has “playlist update” in the help strings for the examples.

Playlist matching message should be changed.
Current: “{number} playlist match {name} Please try to be more specific or use the playlist ID.”
New: “{number} playlists match {name}: Please try to be more specific, or use the playlist ID.

Playlist update is able to be used by anyone, not just the author

Playlist dedupe message should be changed.
Current: “You do not have the permissions to manage playlist owned by {name}.”
New: “You do not have the permissions to manage {playlist name}, owned by {name}.” (Feel free to follow the format for Name/ID in other dialog boxes)
If it’s difficult to pull the playlist name at this point:
New: “You do not have the permissions to manage that playlist, which is owned by {name}.”

"""List all keywords added to the whitelist.""" is doubled.

Bump has some commented lines left in it.

Drapersniper committed:
Address Aika's latest review.

1. Update docstring for `[p]playlist rename`.
2. Fix punctuation for playlist matching.
3. `[p]playlist update` now respect playlist management perms
4. Playlist management errors now shows playlist name, id and scope where possible
5. Remove duplicated code and dead code.

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Pluralize string

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>
